Add IsInnerLoopTestProject and IsOuterLoopTestProject property support during the build

krwq opened the pull request:
Add properties to disable selected projects from Inner- or Outer- loop run. By default all projects are Inner-loop. You can disable it from particular run by explicitly setting TestCategories property

<!---
@huboard:{"order":843.25,"milestone_order":961}
-->

weshaggard commented on src/dir.targets:
```diff
@@ -94,8 +94,18 @@
+
     <IsTestProject Condition="'$(IsTestProject)'=='' And $([System.String]::new('$(MSBuildProjectName)').EndsWith('.Tests'))">True</IsTestProject>
-    <RunTestsForProject Condition="'$(RunTestsForProject)'=='' and '$(IsTestProject)'=='True'">True</RunTestsForProject>
+    <IsTestProject Condition="'$(IsInnerLoopTestProject)'!='True' And '$(IsOuterLoopTestProject)'!='True'">False</IsTestProject>
```
This now requires that ever test project set one or the other? What if they are both and don't care? I don't want to force people to have to set this.

weshaggard replied:
Ignore that I just saw that you defaulted them.

weshaggard commented on src/dir.targets:
```diff
@@ -94,8 +94,18 @@
   <Import Project="$(ToolsDir)publishtest.targets" Condition="Exists('$(ToolsDir)publishtest.targets')" />
 
   <PropertyGroup>
+    <IsInnerLoopTestProject Condition="'$(IsInnerLoopTestProject)'==''">True</IsInnerLoopTestProject>
```
I was hoping that we would have a more general category property. Something like `<TestProjectCategory>InnerLoop;OuterLoop</TestProjectCategory>` and then we could just write conditions against those to do the filtering. I fully expect to get other test categories like Perf, Stress, etc.

krwq replied:
How would you default it? Would it be enforcing to have some category in every test project or rather default to InnerLoop and then:
<TestProjectCategory>$(TestProjectCategory);OuterLoop</TestProjectCategory>

but in this case how would you remove some category?

I feel like it adds additional complexity when we don't need it yet.

weshaggard replied:
I would default it the same way you are doing here. Just set it to be InnerLoop;OuterLoop if it is set to empty.

krwq commented:
@weshaggard, could you look at another iteration?

weshaggard commented on src/dir.targets:
```diff
@@ -157,9 +154,39 @@
   <!-- On command line: run unit tests on CoreCLR as post-build step. -->
   <Target Name="RunTestsForProject"
-          AfterTargets="Test"
+          AfterTargets="CheckTestCategories"
```
You should leave AfterTargets="Test" and add a DependsOnTargets="CheckTestCategories"

weshaggard commented on src/dir.targets:
```diff
@@ -157,9 +154,39 @@
+      <RunInnerLoop Condition="'$(RunInnerLoop)'==''">True</RunInnerLoop>
+      
+      <!-- Run tests only if we want to run tests of the given category and the test has this category -->
+      <RunTestsForProject>False</RunTestsForProject>
```
If RunTestsForProject=false already you should keep it false. That should still be used as a way to disable tests. It is used already by other tests like DriveInfoTests.

krwq replied:
Don't we have "SkipTests" for that?

weshaggard replied:
SkipTests is intended for an entire test/build run not for one project. RunTestsForProject is for just one project.

weshaggard commented on src/dir.targets:
```diff
@@ -157,9 +154,39 @@
+      <!-- Run tests only if we want to run tests of the given category and the test has this category -->
+      <RunTestsForProject>False</RunTestsForProject>
+      <RunTestsForProject Condition="'$(IsInnerLoop)'=='True' And '$(RunInnerLoop)'=='True'">True</RunTestsForProject>
+      <RunTestsForProject Condition="'$(IsOuterLoop)'=='True' And '$(RunOuterLoop)'=='True'">True</RunTestsForProject>
```
You are going to need to move the RunOuterLoop properties into this target as well to make sure it gets included correctly. 

It would kind of be nice of these TestCategories got translated into traits (perhaps we even call the property TestTraits to make it align).

krwq replied:
I'm not sure I understand what you meant by that. Let me send another iteration before fixing it.

nslottow replied:
I think what Wes is saying is that it would be nice if $(TestCategories) were the thing that controls the options passed to Xunit, instead of having individual properties like $(IsInnerLoop) for each category.  That way if we add a new category it just works.

weshaggard replied:
Yes that is along the lines of what I wanted. I could imagine you can actually simplify this do something like:

```
 <XunitOptions>$(XunitOptions) @(TestCategories -> '-trait category=%(Identity)')</XunitOptions>
```

weshaggard commented on src/dir.targets:
```diff
@@ -157,9 +154,39 @@
+          Condition="'$(IsTestProject)'=='True' And '$(SkipTests)'!='True'">
+    <PropertyGroup>
+      <!-- By default project has inner- and outer- loop tests. -->
+      <TestCategories Condition="'$(TestCategories)'==''">InnerLoop;OuterLoop</TestCategories>
```
Perhaps we should just default this to InnerLoop given that InnerLoop will be included when we do an OuterLoop test run.

krwq replied:
I feel like this PR was a bad idea from the start. Now when someone creates OuterLoop tests in a project they would also need to know that you have to change a csproj and add a category. I don't expect people to remember that as I would probably forget about this after few months even if I was the one who implemented it. I would rather wait until new version of Xunit (they already fixed they bug) and keep only traits. I find that having separate test project categories and test categories is an unnecessary effort which leads to eventually saving few seconds during the build. I'd rather wait that time and not have to remember to add anything in your csproj.

weshaggard replied:
We are going to very quickly get to a place where we need to have quick efficient ways of filtering out entire test projects and not always blindly running xUnit to let it determine there isn't anything to do. Right now we are only doing simple csc compiling but when we start introducing .NET Native test runs the cost of just compiling a test project to determine there isn't anything to do will become too high to pay. Perhaps a Roslyn based tool once we move to using Roslyn could be used to help default these values but I do believe we are going to need to have a way to specify them in the project file itself.

krwq commented:
@weshaggard, I just updated this PR. I didn't change AfterTargets as the Conditions are being evaluated after target "Test" is finished and because of that I couldn't disable one target from another. Also I didn't understand one of the comments when you asked me to move some properties to the target so I didn't do that either.

krwq commented:
I need to update that PR as I forgot to undo SkipTests changes from the first commit.

krwq commented:
Ok, should be good now.

krwq commented:
I resolved merge confilcts

krwq commented:
^ Removed whitespaces diffs

weshaggard commented on src/dir.targets:
```diff
@@ -116,11 +116,15 @@
+  <PropertyGroup>
+    <RunInnerLoop Condition="'$(RunInnerLoop)' == ''">True</RunInnerLoop>
+    <RunOuterLoop Condition="'$(RunOuterLoop)' == ''">False</RunOuterLoop>
+  </PropertyGroup>
```
Can this just become RunTestWithTrait (or Category)?

weshaggard commented on src/dir.targets:
```diff
@@ -160,8 +147,35 @@
+    <PropertyGroup Condition="'$(RunInnerLoop)'!='True'">
+      <!-- When inner-loop tests are disabled we want to include categories -->
+      <XunitOptions Condition="'$(RunOuterLoop)'=='True'">$(XunitOptions) -trait category=outerloop</XunitOptions>
+    </PropertyGroup>
```
Should just do something like:
```
 <XunitOptions>$(XunitOptions) @(TestCategories -> '-trait category=%(Identity)')</XunitOptions>
```

That would probably allow you to remove most of the Is/RunInnerLoop properties. And you can filter the TestTraits/Categories via the RunTestTraits item.

weshaggard commented on src/dir.targets:
```diff
@@ -174,6 +188,33 @@
+      <!-- By default project has only inner-loop tests. -->
+      <TestCategories Condition="'$(TestCategories)'==''">InnerLoop</TestCategories>
+    </PropertyGroup>
+    <ItemGroup>
```
This target can simply become a filtering of the TestCategories via the RunTestWithTraits item if you switch to that model.

weshaggard commented on src/dir.targets:
```diff
@@ -160,8 +147,35 @@
   <Target Name="RunTestsForProject"
-          AfterTargets="Test"
-          Condition="'$(RunTestsForProject)'=='true'">
+          AfterTargets="CheckTestCategories"
```
Please add back AfterTargets="Test" and add DependsOnTargets="CheckTestCaegories"

weshaggard commented on src/dir.targets:
```diff
@@ -174,6 +188,33 @@
   </Target>
 
+  <Target Name="CheckTestCategories"
+          AfterTargets="Test"
```
Please remove AfterTargets="Test" and simply make this target a dependency of only RunTestsForProject.

krwq replied:
That doesn't work as Condition is being calculated when Test target is finished and thus I can't change the condition in the target it depends on :( (looks like msbuild bug to me)

weshaggard replied:
Hmmm... OK I will try to see if there is a better way to organize these as I generally prefer a model where targets explicitly state what the depend on instead of creating this fragile target chain that is hard to plug into. For now I'm fine with what you have.

nslottow commented:
LGTM assuming Wes will look into if there's a cleaner way to do this.

weshaggard commented on src/dir.targets:
```diff
@@ -178,6 +203,28 @@
+    <ItemGroup>
+      <!-- Intersection of TestCategoriesItems and RunTestsWithCategoriesItems -->
+      <RunTestsCategoriesFiltered Include="@(RunTestsWithCategoriesItems)" Condition="'@(RunTestsWithCategoriesItems)'=='@(TestCategoriesItems)' And '%(Identity)'!=''" />
+      <RunTestsCategoriesFiltered Remove="@(RunTestsWithoutCategoriesItems)" />
```
Condition="'@(RunTestsWithCategoriesItems)'=='@(TestCategoriesItems)'
I'm not sure I understand this condition

krwq replied:
We want an intersection here. If I remove it then I will get subtract(A, B) instead of intersect(A, B).

This condition does count(A)*count(B) operation and is being expanded for each combination and if element exists in both sets then is being added to the resulting set. This only works because of "'%(Identity)'!=''" which causes it to enumerate. Without that it would simply do:
'JOIN(A,";")'=='JOIN(B,";")' which would not work. In addition to that it removes empty elements which we want too.

weshaggard replied:
As we chatted can you switch this to using "%(...)" in the condition so it is a little clearer what it is trying to do.

weshaggard commented on src/dir.targets:
```diff
@@ -164,8 +158,39 @@
+    </PropertyGroup>
+
+    <!-- xUnit command line without coverage enabled -->
+    <PropertyGroup>
```
Do these need to be in the target? Can we just add more XUnitOptions at the end of the command line?

krwq replied:
I wanted to do that in the beginning but this condition: "'@(RunTestsWithCategoriesItems)'!=''" is required to be in target

krwq commented:
@dotnet-bot test this please

bartonjs commented:
Given that xunit has made a change (https://github.com/xunit/xunit/commit/04bdb67a6bbd12322ffd27f06782d2c7b3e971ba) to not fail when zero cases match the constraint, is there anything that you would want to do differently?

What commandline do we need to use for the CI system with this change?  I'm assuming "Build.cmd /p:RunOuterLoop=True /p:RunTestsWithCategories=OuterLoop", but I think that will exclude the inner loop tests on the upgraded runs (currently it runs both because "RunOuterLoop" being set prevents the exclusion filter from being applied).

I'm not necessarily opposed to losing InnerLoop-classified tests during the OuterLoop execution, but want to make sure that we're aware of that change coming as a result of this (or, if we don't want the change, determining how to avoid it). For data, I feel like I originally wanted it to be OuterLoop only, but the xunit-zero-tests-means-fail feature made it hard.

krwq commented:
This is just generalized way of disabling testing of projects when there are no tests (to reduce overhead of discovering tests etc) - it matches 1:1 with what Xunit traits.
Ideally we would like that noone would have to add test categories manually but that is what we will have right now. Once we upgrade our fork of xunit or use the official one we can make the default option for test categories broader so it will have to be used only when someone tries to create some less popular test categories like Performance, Stress.

To run both inner and outer loop use:
/p:RunTestsWithoutCategories=None

To run only outer loop:
/p:RunTestsWithCategories=OuterLoop

bartonjs commented:
cd src\System.Xml.XmlDocument\tests
msbuild /t:BuildAndTest  (463 tests run)
msbuild /t:BuildAndTest /p:RunTestsWithoutCategories=None (465 tests run)
msbuild /t:BuildAndTest /p:RunTestsWithoutCategories="None;OuterLoop" (463 tests run)
msbuild /t:BuildAndTest /p:RunTestsWithCategories=PotatoFamine (xunit isn't launched)
msbuild /t:BuildAndTest /p:RunTestsWithCategories=OuterLoop (ERROR: xunit isn't launched)
msbuild /t:BuildAndTest /p:RunTestsWithCategories=InnerLoop (463 tests run)
msbuild /t:BuildAndTest /p:RunTestsWithCategories="PotatoFamine;InnerLoop" (ERROR: xunit is launched with -trait category=PotatoFamine (which coincidentally fails due to 0 tests matching))
msbuild /t:BuildAndTest /p:RunTestsWithCategories="InnerLoop;PotatoFamine" (ERROR: xunit is launched with -trait category=PotatoFamine (wanted to see if InnerLoop being first mattered))

So the error of "Running InnerLoop tests with other test categories is currently not supported! To run all tests use: RunTestsWithoutCategories=None" didn't fire, and specifying OuterLoop as the only category seems to be the same as tests being disabled.

krwq commented:
I fixed those scenarios.

To run all tests:
msbuild /t:BuildAndTest /p:RunTestsWithoutCategories=None
(None is a fake category not used anywhere)

To run Inner and Outer loop:
msbuild /t:BuildAndTest /p:RunTestsWithCategories="InnerLoop;OuterLoop"

To run only InnerLoop:
msbuild /t:BuildAndTest
OR
msbuild /t:BuildAndTest /p:RunTestsWithCategories=InnerLoop

To run only OuterLoop:
msbuild /t:BuildAndTest /p:RunTestsWithCategories=OuterLoop

Also: You can mix InnerLoop with other categories now.

The only thing which is forbidden is specifying both RunTestsWithCategories and RunTestsWithoutCategories. It is too confusing to do what you want to do when mixing those categories and thus it is forbidden.

krwq commented:
@dotnet-bot test this please

bartonjs commented:
LGTM, all of my previous cases now do what I'd expect.

weshaggard commented on src/dir.targets:
```diff
@@ -173,8 +166,31 @@
+      <TestHost>$(CoverageHost)</TestHost>
+      <XunitOptions>$(XunitOptions) -parallel none</XunitOptions>
+      <TestCommandLine>$(CoverageCommandLine) -targetargs:"$(TestCommandLine)"</TestCommandLine>
+    </PropertyGroup>
```
Is there a reason we need to have all these properties inside the target? Can we simply append the extra computed xunit options to the existing options?

krwq replied:
I was modifying previous version so I'm not sure. I'll recheck if it is still required to be there

krwq replied:
I was able to pull out some of the properties but not all of them. We modify XunitOptions in the target so anything depending on XunitOptions needs to stay

weshaggard commented on src/dir.targets:
```diff
@@ -187,6 +203,36 @@
+          Condition="'$(TestDisabled)'!='True'">
+
+    <Error Condition="'$(RunTestsWithCategories)' != '' And '$(RunTestsWithoutCategories)' != ''"
+           Text="Specifying both RunTestsWithCategories and RunTestsWithoutCategories is ambiguous." />
```
Change this to:
Specifying both RunTestsWithCategories and RunTestsWithoutCategories is not supported please only specify one or the other.

krwq replied:
Will do.

weshaggard commented on src/dir.targets:
```diff
@@ -187,6 +203,36 @@
+      <!-- Intersection of TestCategoriesItems and RunTestsWithCategoriesItems -->
+      <TestCategoriesToRun Include="@(TestCategoriesItems)" Condition="'@(RunTestsWithCategoriesItems)'=='%(Identity)'" />
+    </ItemGroup>
+    <ItemGroup Condition="'@(RunTestsWithoutCategoriesItems)'!=''">
```
I would combine both of these item groups given they have the same condition. It is good to eliminate the duplication.

weshaggard commented on src/dir.targets:
```diff
@@ -187,6 +203,36 @@
+    <ItemGroup Condition="'@(RunTestsWithoutCategoriesItems)'!=''">
+      <!-- Run all test categories except specified -->
+      <TestCategoriesToRun Include="@(TestCategoriesItems)" />
+      <TestCategoriesToRun Remove="@(RunTestsWithoutCategoriesItems)" />
```
Remove works but for future you can also use Exclude on the same item group you did the Include on.

krwq replied:
I didn't know about that :) I'll test if it works as expected and modify if works good.

weshaggard commented:
Thanks for your work on this @krwq I've added a couple more minor comments please look at those small tweaks but beyond that it looks good.

bartonjs commented:
I've updated the dotnet_corefx_outerloop job to pass both /p:RunOuterLoop=True (for current/back-compat) and /p:RunTestsWithCategories="InnerLoop;OuterLoop".  So everything should Just Work once this gets merged.

krwq commented:
I addressed the feedback:
- moved properties not depending on XunitOptions outside of target
- replaced Include&Remove with single Include & Exclude
- changed comment

krwq commented:
@bartonjs, do you want to add anything to this PR before merge? (i.e. integrate with CI) I'm not sure if anything is depending on the msbuild properties

weshaggard commented on src/dir.targets:
```diff
@@ -192,6 +208,35 @@
+      <RunWithTraits Condition="'@(RunWithInnerLoopItems)'==''" Include="@(TestCategoriesToRun)" />
+
+      <RunWithoutTraits Condition="'@(RunWithInnerLoopItems)'!=''" Include="@(TestCategoriesItems)" />
+      <RunWithoutTraits Condition="'@(RunWithInnerLoopItems)'!=''" Remove="@(TestCategoriesToRun)" />
```
nit: This Remove could also be an Exclude.

weshaggard commented:
LGTM Thanks @krwq

bartonjs commented:
Nope, in theory the CI support is already done.  I'm good with the merge.
